<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Peer Assignment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:
[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:
[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:
[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. Click on the links below to download and save the datasets (.CSV files):

*   <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Chicago Census Data</a>

*   <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Chicago Public Schools</a>

*   <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Chicago Crime Data</a>

**NOTE:** Ensure you have downloaded the datasets using the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**

Let us now load the ipython-sql  extension and establish a connection with the database

* Here you will be loading the csv files into the pandas Dataframe and then loading the data into the above mentioned sqlite tables.

* Next you will be connecting to the sqlite database  **FinalDB**.

Refer to the previous lab for hints .

<a href ="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Module%205/DB0201EN-Week3-1-4-Analyzing_SQLite.ipynb">Hands-on Lab: Analyzing a real World Data Set</a>





In [1]:
import sqlite3 as sql
import pandas as pd
from pathlib import Path


Path("../data/ChicagoData.db").touch()
con = sql.connect("../data/ChicagoData.db")
c = con.cursor()

In [3]:
census_data = pd.read_csv("../data/ChicagoCensusData.csv")
crime_data = pd.read_csv("../data/ChicagoCrimeData.csv")
school_data = pd.read_csv("../data/ChicagoPublicSchools.csv")

In [4]:
c.execute("""CREATE TABLE CensusData (COMMUNITY_AREA_NUMBER int, COMMUNITY_AREA_NAME int, PERCENT_OF_HOUSING_CROWDED float, PERCENT_HOUSEHOLDS_BELOW_POVERTY float, PERCENT_AGED_16__UNEMPLOYED float, PERCENT_AGED_25__WITHOUT_HIGH_SCHOOL_DIPLOMA float, PERCENT_AGED_UNDER_18_OR_OVER_64 float, PER_CAPITA_INCOME int, HARDSHIP_INDEX int)""")

OperationalError: table CensusData already exists

In [5]:
census_data.to_sql("CensusData", con, if_exists='replace', index=False)

78

In [6]:
c.execute("SELECT * FROM CensusData").fetchall()

[(1, 'Rogers Park', 7.7, 23.6, 8.7, 18.2, 27.5, 23939, 39),
 (2, 'West Ridge', 7.8, 17.2, 8.8, 20.8, 38.5, 23040, 46),
 (3, 'Uptown', 3.8, 24.0, 8.9, 11.8, 22.2, 35787, 20),
 (4, 'Lincoln Square', 3.4, 10.9, 8.2, 13.4, 25.5, 37524, 17),
 (5, 'North Center', 0.3, 7.5, 5.2, 4.5, 26.2, 57123, 6),
 (6, 'Lake View', 1.1, 11.4, 4.7, 2.6, 17.0, 60058, 5),
 (7, 'Lincoln Park', 0.8, 12.3, 5.1, 3.6, 21.5, 71551, 2),
 (8, 'Near North Side', 1.9, 12.9, 7.0, 2.5, 22.6, 88669, 1),
 (9, 'Edison Park', 1.1, 3.3, 6.5, 7.4, 35.3, 40959, 8),
 (10, 'Norwood Park', 2.0, 5.4, 9.0, 11.5, 39.5, 32875, 21),
 (11, 'Jefferson Park', 2.7, 8.6, 12.4, 13.4, 35.5, 27751, 25),
 (12, 'Forest Glen', 1.1, 7.5, 6.8, 4.9, 40.5, 44164, 11),
 (13, 'North Park', 3.9, 13.2, 9.9, 14.4, 39.0, 26576, 33),
 (14, 'Albany Park', 11.3, 19.2, 10.0, 32.9, 32.0, 21323, 53),
 (15, 'Portage Park', 4.1, 11.6, 12.6, 19.3, 34.0, 24336, 35),
 (16, 'Irving Park', 6.3, 13.1, 10.0, 22.4, 31.6, 27249, 34),
 (17, 'Dunning', 5.2, 10.6, 10.0, 16.2,

In [7]:
c.execute("""CREATE TABLE CrimeData (ID int, CASE_NUMBER text, DATE date, BLOCK text, IUCR int, PRIMARY_TYPE text, DESCRIPTION text, LOCATION_DESCRIPTION text, ARREST bool, DOMESTIC bool, BEAT int, DISTRICT int, WARD int, COMMUNITY_AREA_NUMBER int, FBICODE int, X_COORDINATE int, Y_COORDINATE int, YEAR int, LATITUDE real, LONGITUDE real, LOCATION text)""")

In [10]:
crime_data.to_sql("CrimeData", con, if_exists='append', index=False)

533

In [11]:
c.execute("SELECT * FROM CrimeData").fetchall()

[(3512276,
  'HK587712',
  '2004-08-28',
  '047XX S KEDZIE AVE',
  890,
  'THEFT',
  'FROM BUILDING',
  'SMALL RETAIL STORE',
  0,
  0,
  911,
  9,
  14,
  58,
  6,
  1155838,
  1873050,
  2004,
  41.8074405,
  -87.70395585,
  '(41.8074405, -87.703955849)'),
 (3406613,
  'HK456306',
  '2004-06-26',
  '009XX N CENTRAL PARK AVE',
  820,
  'THEFT',
  '$500 AND UNDER',
  'OTHER',
  0,
  0,
  1112,
  11,
  27,
  23,
  6,
  1152206,
  1906127,
  2004,
  41.89827996,
  -87.71640551,
  '(41.898279962, -87.716405505)'),
 (8002131,
  'HT233595',
  '2011-04-04',
  '043XX S WABASH AVE',
  820,
  'THEFT',
  '$500 AND UNDER',
  'NURSING HOME/RETIREMENT HOME',
  0,
  0,
  221,
  2,
  3,
  38,
  6,
  1177436,
  1876313,
  2011,
  41.81593313,
  -87.62464213,
  '(41.815933131, -87.624642127)'),
 (7903289,
  'HT133522',
  '2010-12-30',
  '083XX S KINGSTON AVE',
  840,
  'THEFT',
  'FINANCIAL ID THEFT: OVER $300',
  'RESIDENCE',
  0,
  0,
  423,
  4,
  7,
  46,
  6,
  1194622,
  1850125,
  2010,
  41.743

In [12]:
school_data.to_sql("SchoolData", con, if_exists="append", index=False)

566

In [13]:
c.execute("SELECT * FROM SchoolData").fetchall()

[(610038,
  'Abraham Lincoln Elementary School',
  'ES',
  '615 W Kemper Pl',
  'Chicago',
  'IL',
  60614,
  '(773) 534-5720',
  'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610038.pdf',
  'Fullerton Elementary Network',
  'NORTH-NORTHWEST SIDE COLLABORATIVE',
  'No',
  'Standard',
  'Not on Probation',
  'Level 1',
  'Yes',
  'Very Strong',
  99.0,
  'Very Strong',
  '99',
  'Strong',
  74.0,
  'Strong',
  66.0,
  'Weak',
  '65',
  'Strong',
  '70',
  'Strong',
  '56',
  'Average',
  '47',
  '96.00%',
  2.0,
  '96.40%',
  '95.80%',
  '80.1',
  '43.3',
  '89.6',
  '84.9',
  '60.7',
  '62.6',
  '81.9',
  '85.2',
  '52',
  '62.4',
  '66.3',
  '77.9',
  69.7,
  64.4,
  0.2,
  0.9,
  'Yellow',
  'Green',
  '67.1',
  '54.5',
  'NDA',
  'NDA',
  'NDA',
  'NDA',
  'NDA',
  'NDA',
  'NDA',
  'NDA',
  'NDA',
  'NDA',
  813,
  33,
  'NDA',
  1171699.458,
  1915829.428,
  41.92449696,
  -87.64452163,
  7,
  'LINCOLN PARK',
  43,
  18,
  '(41.92449696, -87.64452163)'),
 (6

## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.

In [64]:
pd.read_sql("SELECT COUNT(*) FROM CrimeData", con)

COUNT(*)
0       533

### Problem 2

##### List community areas with per capita income less than 11000.

In [17]:
pd.read_sql("SELECT COMMUNITY_AREA_NAME FROM CensusData WHERE PER_CAPITA_INCOME < 11000", con)

COMMUNITY_AREA_NAME
0  West Garfield Park
1      South Lawndale
2         Fuller Park
3           Riverdale

### Problem 3

##### List all case numbers for crimes  involving minors?(children are not considered minors for the purposes of crime analysis)


In [16]:
pd.read_sql("SELECT CASE_NUMBER FROM CrimeData WHERE DESCRIPTION LIKE \"%MINOR%\"", con)

CASE_NUMBER
0    HL266884
1    HK238408

### Problem 4

##### List all kidnapping crimes involving a child?

In [15]:
pd.read_sql("SELECT * FROM CrimeData WHERE DESCRIPTION LIKE \"%CHILD ABDUCTION%\"", con)

ID CASE_NUMBER        DATE                 BLOCK  IUCR PRIMARY_TYPE  \
0  5276766    HN144152  2007-01-26  050XX W VAN BUREN ST  1792   KIDNAPPING   

                DESCRIPTION LOCATION_DESCRIPTION  ARREST  DOMESTIC  ...  \
0  CHILD ABDUCTION/STRANGER               STREET       0         0  ...   

   DISTRICT  WARD  COMMUNITY_AREA_NUMBER  FBICODE  X_COORDINATE  Y_COORDINATE  \
0        15    29                     25       20       1143050       1897546   

   YEAR   LATITUDE  LONGITUDE                       LOCATION  
0  2007  41.874908 -87.750249  (41.874908413, -87.750249307)  

[1 rows x 21 columns]

### Problem 5

##### What kinds of crimes were recorded at schools?


In [67]:
pd.read_sql("SELECT PRIMARY_TYPE, COUNT(PRIMARY_TYPE) FROM CrimeData WHERE LOCATION_DESCRIPTION LIKE \"%SCHOOL%\" GROUP BY PRIMARY_TYPE", con)

PRIMARY_TYPE  COUNT(PRIMARY_TYPE)
0                 ASSAULT                    1
1                 BATTERY                    5
2         CRIMINAL DAMAGE                    1
3       CRIMINAL TRESPASS                    1
4               NARCOTICS                    2
5  PUBLIC PEACE VIOLATION                    2

### Problem 6

##### List the average safety score for each type of school.

In [23]:
pd.read_sql("SELECT \"Elementary, Middle, or High School\", Avg(SAFETY_SCORE) FROM SchoolData GROUP BY \"Elementary, Middle, or High School\"", con)

Elementary, Middle, or High School  Avg(SAFETY_SCORE)
0                                 ES          49.520384
1                                 HS          49.623529
2                                 MS          48.000000

### Problem 7

##### List 5 community areas with highest % of households below poverty line

In [14]:
result = pd.read_sql("SELECT COMMUNITY_AREA_NUMBER, COMMUNITY_AREA_NAME, PERCENT_HOUSEHOLDS_BELOW_POVERTY FROM CensusData", con)
result.sort_values("PERCENT_HOUSEHOLDS_BELOW_POVERTY", ascending=False)[:5]

COMMUNITY_AREA_NUMBER COMMUNITY_AREA_NAME  \
53                   54.0           Riverdale   
36                   37.0         Fuller Park   
67                   68.0           Englewood   
28                   29.0      North Lawndale   
26                   27.0  East Garfield Park   

    PERCENT_HOUSEHOLDS_BELOW_POVERTY  
53                              56.5  
36                              51.2  
67                              46.6  
28                              43.1  
26                              42.4

### Problem 8

##### Which community area is most crime prone?

In [46]:
query1 = "SELECT COMMUNITY_AREA_NUMBER, COUNT(COMMUNITY_AREA_NUMBER) FROM CrimeData GROUP BY COMMUNITY_AREA_NUMBER"
r = pd.read_sql(query1, con).dropna()
r.sort_values("COUNT(COMMUNITY_AREA_NUMBER)", inplace=True, ascending=False)

query2 = f"""
SELECT COMMUNITY_AREA_NUMBER, COMMUNITY_AREA_NAME
FROM CensusData
WHERE COMMUNITY_AREA_NUMBER = {r.iloc[0]['COMMUNITY_AREA_NUMBER']}"""
r2 = pd.read_sql(query2, con)

r2


COMMUNITY_AREA_NUMBER COMMUNITY_AREA_NAME
0                     25              Austin

Double-click **here** for a hint

<!--
Query for the 'community area number' that is most crime prone.
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


In [48]:
query = "SELECT COMMUNITY_AREA_NAME, HARDSHIP_INDEX FROM CensusData WHERE HARDSHIP_INDEX = (SELECT MAX(HARDSHIP_INDEX) FROM CensusData)"
pd.read_sql(query, con)

COMMUNITY_AREA_NAME  HARDSHIP_INDEX
0           Riverdale              98

### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


In [63]:
query = f"""
SELECT CR.COMMUNITY_AREA_NUMBER, CD.COMMUNITY_AREA_NAME, COUNT(CR.ID) AS CRIME_NO
FROM CrimeData AS CR, CensusData AS CD
WHERE CR.COMMUNITY_AREA_NUMBER = CD.COMMUNITY_AREA_NUMBER
GROUP BY CR.COMMUNITY_AREA_NUMBER,  CD.COMMUNITY_AREA_NAME
ORDER BY CRIME_NO DESC LIMIT 1"""

pd.read_sql(query, con)


COMMUNITY_AREA_NUMBER COMMUNITY_AREA_NAME  CRIME_NO
0                     25              Austin        43

Copyright © 2020 This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ).


## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>

## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
